### PR TITLE
Desktop: whatsapp integration

### PR DESCRIPTION
**Pull request by PriyanshuDangi**
## Summary

This PR adds WhatsApp support to Omi Desktop so users can connect WhatsApp, view their chats, and let Omi help draft or send replies in the user's own tone.

Main features:

- Connect WhatsApp from the desktop app by scanning a QR code.
- Add a new Messages page for viewing WhatsApp chats and messages.
- Let Omi read recent WhatsApp context and generate reply drafts.
- Support approve, dismiss, and always-auto-reply flows for generated drafts.
- Add optional auto-reply mode with guardrails like allowlists, kill switch, rate limits, quiet hours, sensitive-content checks, and audit logs.
- Build a lightweight tone profile from the user's sent WhatsApp messages so replies sound closer to how the user writes.
- Optionally sync WhatsApp history into Omi memories only when the user enables WhatsApp sync.
- Expose WhatsApp tools to the desktop agent through the existing Omi bridge, with guarded manual sending.

## Why `@openclaw/wacli`

The goal is to make Omi act more like an AI clone that can understand the user's messaging style and reply on their behalf when allowed. For that, Omi needs access to WhatsApp chats, recent conversation context, and the user's own sent messages for tone matching.

`@openclaw/wacli` gives us a simple local way to do this on desktop: the user just scans a WhatsApp QR code, like linking a normal WhatsApp device. We do not need to build backend WhatsApp infrastructure for this flow, and the WhatsApp session stays local to the desktop app.

## Design Notes

- WhatsApp connection, sync, reading, and sending are handled locally by the Swift desktop app through `wacli`.
- The new Messages UI uses a generic provider layer so future messaging apps can be added later.
- Reply automation is draft-first by default. Auto-send requires explicit settings and allowlisting.
- Memory import is opt-in, so WhatsApp messages are only added to Omi memories when the user enables sync.
- The agent uses Swift-backed WhatsApp tools through the existing desktop bridge so the same user settings and safety checks apply everywhere.

## Review Notes

This is a privacy-sensitive feature because it touches WhatsApp messages, contacts, memory sync, and reply sending. The main review areas are product/privacy approval, send/auto-reply guardrails, and the packaging/provenance decision for the WhatsApp CLI dependency.

<a href="https://cubic.dev/pr/BasedHardware/omi/pull/8462?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

**Review by Git-on-my-level — CHANGES_REQUESTED** [review]
Thanks for the contribution. I’m going to request changes rather than approve this because the PR introduces a broad, security/privacy-sensitive WhatsApp integration and bundles a new executable.

Blocking concerns:

- `desktop/macos/Desktop/Sources/Resources/wacli` is a ~44 MB Mach-O universal executable added directly to the PR. I can’t treat an opaque binary from a first-time contributor as reviewable or safe to ship. Please remove the bundled binary from the PR and instead document or implement a trusted, reproducible supply-chain path for obtaining/building it, including version/source/provenance and maintainer-controlled release packaging.
- The feature gives the desktop app access to WhatsApp messages/contacts, can import private messages into Omi memories, starts a local webhook, exposes WhatsApp tools to the agent, and includes message-sending/auto-reply flows. That needs explicit maintainer product/security/privacy review before merge.
- The change is very large and difficult to review as one unit (35 files / ~6k changed lines), with only a very small test addition. Please split this into smaller PRs, e.g. UI shell, wacli/runtime packaging, read-only message sync, memory import, and send/auto-reply guardrails, with focused tests for each.
- For any future version that can send or auto-send messages, please add tests and documentation around consent, allowlisting, kill switch behavior, rate limits, quiet hours, audit logs, and prompt/tool-call boundaries.

This is not a rejection of the idea; WhatsApp could be valuable for Omi. But this specific PR needs a much narrower, reviewable shape and maintainer-owned security/provenance decisions before it can proceed.

**Review by Git-on-my-level — CHANGES_REQUESTED** [review]
Thanks for continuing to iterate on this. I’m still requesting changes on the current head (`ce8e263`) rather than approving.

The main blockers remain:

- `desktop/macos/Desktop/Sources/Resources/wacli` is still a bundled ~44 MB Mach-O executable (`3667afdad78021daa6825780208b8e2459701a91582479d4ca18e27248ceed2c`). An opaque binary from a first-time contributor is not reviewable enough for this repo. Please remove it from the PR and use a maintainer-controlled, reproducible/provenance-backed packaging path for any WhatsApp runtime dependency.
- This PR adds a high-risk surface: WhatsApp message/contact access, importing private WhatsApp history into Omi memories, a local webhook, agent-exposed WhatsApp tools, direct sending, and auto-reply/draft flows. That needs explicit product, privacy, and security review before merge.
- The change is still too broad for safe review as one unit (38 files / ~6.6k changed lines) and has very limited tests relative to the risk. Please split this into narrower PRs such as runtime/provenance, UI shell, read-only sync, memory import, and send/auto-reply guardrails.
- Before any send or auto-reply functionality can be considered, it needs focused tests/docs around consent, allowlisting, kill switch behavior, rate limits/quiet hours, audit logs, prompt/tool-call boundaries, and wrong-recipient prevention.

The idea may be valuable for Omi, but this needs a much narrower, source-reviewable shape plus maintainer-owned security/provenance decisions before it can proceed.

**Comment by PriyanshuDangi** [comment]
@Git-on-my-level thanks for the feedback

1. I will add more tests and add a doc or summary explaining all the design and decisions
2. I'll resolve the cubic comments
3. I am working on the AI clone feature in which the requirement is 
`I want omi Omi to respond to people on my behalf using my apps`
And for this I am using @openclaw/wacli using which people can just login with a QR and can simply have draft and auto reply features 
(And it already syncs in with the brain only if the user clicks on the allow sync for whatsapp)

**Review by Git-on-my-level — CHANGES_REQUESTED** [review]
Thanks for the continued iteration. I’m requesting changes on the current head (`de83fa2`) rather than approving.

The existing high-level blockers still apply: this is a broad WhatsApp integration touching private message/contact access, memory import, local webhook handling, agent-exposed WhatsApp tools, direct sending, auto-reply, runtime packaging, and an opaque bundled `wacli` executable. It needs maintainer product/security/privacy/provenance review and a much smaller split before merge.

One specific blocker I want to call out on this head:

- `WhatsAppReplyCoordinator.draftReply` exposes the full `ChatToolExecutor.execute` tool surface to a model session whose prompt includes untrusted WhatsApp message text and recent thread content. Because that tool surface includes `wa_send_message`, a malicious incoming WhatsApp message could try to prompt-inject the drafting agent into sending messages during draft generation, before the UI approval/allowlist/auto-reply decision path. Draft generation should be read-only by construction: expose only safe read/context tools there, block `wa_send_message` from that session, and route all sends exclusively through explicit user approval or the separately reviewed auto-reply guardrail path.

Additional blockers remain:

- `desktop/macos/Desktop/Sources/Resources/wacli` is still a bundled ~44 MB Mach-O universal executable (`3667afdad78021daa6825780208b8e2459701a91582479d4ca18e27248ceed2c`). Please remove opaque binaries from the PR and use a maintainer-controlled, reproducible/provenance-backed packaging path.
- The PR is still too large for safe review as one unit (43 files / ~6.7k changed lines) with very limited tests relative to the risk. Please split into narrower PRs, e.g. runtime/provenance, UI shell, read-only sync, memory import, and send/auto-reply guardrails.
- Send/auto-reply functionality needs focused tests/docs for consent, prompt/tool-call boundaries, wrong-recipient prevention, allowlisting, kill switch behavior, rate limits/quiet hours, audit logs, and failure/retry semantics.

The product idea may be valuable for Omi, but this needs a much narrower and source-reviewable shape before it can proceed.

**Review by Git-on-my-level — CHANGES_REQUESTED** [review]
Thanks for adding more tests and continuing to work through the review feedback. I’m still requesting changes on the current head (`919cea4`) rather than approving.

The new test coverage is a step in the right direction, but it does not resolve the main merge blockers for this PR:

- This is still a very broad WhatsApp integration touching private messages/contacts, memory import, local webhook handling, agent-exposed WhatsApp tools, direct sending, auto-reply, runtime packaging, and desktop workflow scripts. That needs human maintainer product/security/privacy/provenance review and a smaller split before merge.
- `desktop/macos/Desktop/Sources/Resources/wacli` is still an opaque bundled ~44 MB Mach-O executable (`3667afdad78021daa6825780208b8e2459701a91582479d4ca18e27248ceed2c`). Please remove the binary from the PR and use a maintainer-owned, source-reviewable/reproducible provenance path for the WhatsApp runtime.
- The previous draft-generation boundary issue is still present: `WhatsAppReplyCoordinator.draftReply` forwards every model tool call to `ChatToolExecutor.execute`, whose surface includes `wa_send_message`, while the prompt includes untrusted WhatsApp message/thread text. Draft generation should be read-only by construction; sending should only happen through explicit user approval or the separately reviewed auto-reply path.
- The latest testability change makes `WhatsAppReplySettings.init(defaults:)` module-visible in production code. That class owns in-memory duplicate-send and auto-send rate-limit state, so allowing multiple production instances can fragment guardrail state. Please keep the singleton guarantee in production and expose a DEBUG/test-only initializer or dependency-injection seam instead.
- `wa-tools-stdio` now generates a send idempotency key from the JSON-RPC request id. If an MCP/client retry uses a different request id for the same logical send, it can produce a different `client_message_id`, weakening duplicate-send protection for the exact path that most needs idempotency. The dedupe key should be stable for the logical send intent/attempt, not merely the transport request id.

The product idea may be valuable, but this PR still needs to be split into narrower, source-reviewable pieces with focused security/privacy tests and maintainer-owned decisions around WhatsApp runtime provenance, message-send boundaries, auto-reply consent/kill-switch/rate-limit behavior, wrong-recipient prevention, and auditability.

**Comment by Git-on-my-level** [comment]
@PriyanshuDangi is there a reason the binary has to ship with the app? What about letting users opt-in and download the required dependency?

**Comment by PriyanshuDangi** [comment]
@Git-on-my-level thanks for checking this - let me make it opt-in to download

**Comment by PriyanshuDangi** [comment]
Screenshot 2026-07-10 at 10 25 55 AM
@Git-on-my-level I have updated it to download on the fly 

And here's a screenshot of how it looks and works

File: desktop/macos/run.sh [hunk 1]
```diff
@@ -530,7 +530,20 @@ if [ -d "$ONNX_FRAMEWORK" ]; then
...
+if [ ! -f "$WEBP_LIB" ]; then
+    for candidate in \
+        /opt/homebrew/opt/webp/lib/libwebp.7.dylib \
+        /usr/local/opt/webp/lib/libwebp.7.dylib; do
+        if [ -f "$candidate" ]; then
+            WEBP_LIB="$candidate"
+            break
+        fi
+    done
+fi
```
Comment (PriyanshuDangi) [inline desktop/macos/run.sh:546]: will revert the changes in this file(once its about to be merged) - it wasn't working without this on my mac

**Review by Git-on-my-level — CHANGES_REQUESTED** [review]
Thanks for removing the bundled `wacli` executable and moving toward opt-in installation. That is a meaningful improvement over shipping the opaque binary in the repo.

I’m still requesting changes on the current head (`aac79bb`) rather than approving, because several merge blockers remain:

- The helper is now downloaded at runtime from `github.com/openclaw/wacli` and installed after only a pinned archive hash check, then quarantine/provenance xattrs are cleared. That removes the repo-size/bundled-binary problem, but it is still a third-party executable supply-chain/provenance decision for a privacy-sensitive WhatsApp feature. This needs a maintainer-owned provenance/update/signing policy before merge.
- The draft-generation boundary is still not read-only by construction. `WhatsAppReplyCoordinator.draftReply` includes untrusted WhatsApp thread/message text in the agent prompt and forwards every tool call through `ChatToolExecutor.execute`; that tool surface includes `wa_send_message`. The prompt asks the model not to send, but the code should enforce a read-only allowlist for draft generation.
- `WhatsAppReplySettings.init(defaults:)` is still module-visible even though the class owns in-memory duplicate-send and auto-send rate-limit state. Please keep the production singleton guarantee and expose a test-only/dependency-injection seam that cannot be accidentally used by production code.
- The generated idempotency key for MCP `wa_send_message` still includes request-scoped IDs (`mcpRequestId`, request/run/attempt fields). A retry of the same logical send through a different request scope can produce a different `client_message_id`, so it does not reliably dedupe the path that most needs dedupe.
- This PR also mixes the WhatsApp desktop integration with unrelated sync/fair-use rate-limit backend/app changes. Those are separate risk surfaces with different reviewers and rollback paths; splitting those unrelated pieces would materially improve review and testing.

The feature direction may be valuable, and the new tests are helpful, but this still needs human maintainer product/security/privacy/dependency review before merge. I did not run the untrusted WhatsApp helper or PR code during this review; this is based on static review of the current diff.

**Review by Git-on-my-level — CHANGES_REQUESTED** [review]
Thanks for the follow-up fixes. The new read-only draft mode in `ChatToolExecutor` and the stable logical idempotency key in `wa-tools-stdio.ts` are meaningful improvements over the last head.

I’m still requesting changes rather than approving because this remains a privacy/security/dependency-sensitive WhatsApp integration and the current head still has merge blockers:

- `desktop/macos/CHANGELOG.json` appears to delete the already-released `0.12.67` entry from `releases` and move this PR’s WhatsApp notes into `unreleased`. Please rebase/restore the released changelog entry and keep this PR to additive unreleased changelog fragments only.
- The app still downloads and executes `openclaw/wacli` at runtime from GitHub, verifies only a pinned archive hash, and then clears quarantine/provenance xattrs after install. That is a product/security/dependency policy decision for a feature that reads contacts/private messages and can send messages. I think this still needs maintainer-owned provenance/signing/update UX (and explicit human review) before merge.
- The feature is still a broad human-judgment surface: local WhatsApp session handling, private message/contact sync, optional brain/memory import, agent-exposed WhatsApp tools, draft generation, allowlisted auto-send, and local webhook processing. The added tests are helpful, but I don’t see a passing CI summary on this head and I did not run the untrusted WhatsApp helper/PR code.
- Some unrelated desktop build/tooling changes are still bundled with the WhatsApp feature (`run.sh`, `prepare-desktop-bundle-native-deps.sh`, pi mono extension schema test/code whitespace). These are separable risk/review surfaces; please move them out unless they are strictly required for the WhatsApp integration.

Human maintainer product/security/privacy/dependency review is still needed before this can merge. I did not approve, merge, or push to the contributor branch.

**Comment by aryanorastar** [comment]
Leaving this open. WhatsApp ingest/send is still missing on main, so this is not a product no.

It is not mergeable on 7120376. The branch is CONFLICTING on DesktopHomeView / DashboardPage / Apps / Settings — the QueryShell shell. CHANGES_REQUESTED on this head still covers the wacli download/execute path, changelog rewrite, and drive-by scripts.

Do not rebase this 8k-line megapack onto current Home. If WhatsApp is still wanted, that is a new shrink PR on today's connector surface, without a second chat timeline and without downloading a helper and clearing quarantine to run it.

**Comment by aryanorastar** [comment]
Keeping this open. Not approving `7120376b1b73`. WhatsApp integration path absent on origin/main. +8270. No check rollup. needs-scope-reduction.

— Astar, on behalf of @aryanorastar
